plugins/experimental/otel_tracer/otel_tracer.cc

```diff
@@ -0,0 +1,363 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+  http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include <ctype.h>
+#include <stdio.h>
+#include <string.h>
+#include <iostream>
+#include <vector>
+#include <unistd.h>
+#include <getopt.h>
+
+#include "ts/ts.h"
+
+#define PLUGIN_NAME "otel_tracer"
+
+#include "tracer_common.h"
+namespace trace    = opentelemetry::trace;
+namespace nostd    = opentelemetry::nostd;
+namespace sdktrace = opentelemetry::sdk::trace;
+
+static int
+close_txn(TSCont contp, TSEvent event, void *edata)
+{
+  int retval = 0;
+  TSMBuffer buf;
+  TSMLoc hdr_loc;
+
+  ExtraRequestData *req_data = static_cast<ExtraRequestData *>(TSContDataGet(contp));
+
+  TSHttpTxn txnp = static_cast<TSHttpTxn>(edata);
+  if (event != TS_EVENT_HTTP_TXN_CLOSE) {
+    TSError("[otel_tracer][%s] Unexpected event (%d)", __FUNCTION__, event);
+    goto lReturn;
+  }
+
+  if (TSHttpTxnClientRespGet(txnp, &buf, &hdr_loc) == TS_SUCCESS) {
+    int status = TSHttpHdrStatusGet(buf, hdr_loc);
+    req_data->span->SetAttribute("http.status_code", status);
+    if (status > 499) {
+      req_data->span->SetStatus(trace::StatusCode::kError);
+    }
+
+    TSHandleMLocRelease(buf, TS_NULL_MLOC, hdr_loc);
+  }
+
+  retval = 1;
+
+lReturn:
+  req_data->Destruct(req_data);
+  delete req_data;
+
+  TSContDestroy(contp);
+  TSHttpTxnReenable(txnp, TS_EVENT_HTTP_CONTINUE);
+
+  return retval;
+}
+
+static void
+set_request_header(TSMBuffer buf, TSMLoc hdr_loc, const char *key, int key_len, const char *val, int val_len)
+{
+  TSMLoc field_loc, tmp;
```

Comment:
Be lazy and declare these as needed within the scope used.  You may be able constify the variables at that point.